### ISSUE TITLE
Linkerd injected apps fails when connecting to pods deployed with hostNetworking. 

### ISSUE THREAD

**Issue by prasus** [linkerd/linkerd2#5938]
### What is the issue?

A couple of our linkerd injected apps are failing when connecting to our non meshed scyllaDB pods, The scyllaDB pods are deployed using `hostNetwork: true`,  it works without any issues when connecting to scyllaDB pods deployed **without** using `hostNetwork: true` option.

I came across [issue/5394](https://github.com/linkerd/linkerd2/issues/5394) and this [comment](https://github.com/linkerd/linkerd2/issues/5394#issuecomment-746882232) explains about this behavior and this supposed to have fixed with [pull/4335](https://github.com/linkerd/linkerd2/pull/4335), but we are seeing the same behavior with `stable-2.10.0` any help on fixing this would be much appreciated, thanks! 

### How can it be reproduced?

1. Deploy [scyllaDB operator](http://operator.docs.scylladb.com/stable/index.html) and then create a scyllaDB cluster with `hostNetwork: true` option.
2. Try connecting to the scylladb service endpoint (non-meshed) from a meshed client pod, this will fail with the below errors in the `linkerd-proxy`

### Logs, error output, etc

```
2 error: protocol error: unexpected internal error encountered", details: [], metadata: MetadataMap { headers: {} }
[   449.069280s]  WARN ThreadId(01) outbound:accept{client.addr=10.244.29.174:60378 target.addr=10.23.83.47:9042}: linkerd_service_profiles::client: Could not fetch profile error=status: Unknown, message: "http2 error: protocol error: unexpected internal error encountered", details: [], metadata: MetadataMap { headers: {} }
[   449.069299s]  WARN ThreadId(01) outbound:accept{client.addr=10.244.29.174:39982 target.addr=10.23.83.47:9042}: linkerd_service_profiles::client: Could not fetch profile error=status: Unknown, message: "http2 error: protocol error: unexpected internal error encountered", details: [], metadata: MetadataMap { headers: {} }
[   449.069324s]  WARN ThreadId(01) outbound:accept{client.addr=10.244.29.174:51980 target.addr=10.23.83.47:9042}: linkerd_service_profiles::client: Could not fetch profile error=status: Unknown, message: "http2 error: protocol error: unexpected internal error encountered", details: [], metadata: MetadataMap { headers: {} }
[   449.069346s]  WARN ThreadId(01) outbound:accept{client.addr=10.244.29.174:59132 target.addr=10.23.83.34:9042}: linkerd_service_profiles::client: Could not fetch profile error=status: Unknown, message: "http2 error: protocol error: unexpected internal error encountered", details: [], metadata: MetadataMap { headers: {} }
```
#### `linkerd check` output

```text
kubernetes-api
--------------
√ can initialize the client
√ can query the Kubernetes API

kubernetes-version
------------------
√ is running the minimum Kubernetes API version
√ is running the minimum kubectl version

linkerd-existence
-----------------
√ 'linkerd-config' config map exists
√ heartbeat ServiceAccount exist
√ control plane replica sets are ready
√ no unschedulable pods
√ controller pod is running

linkerd-config
--------------
√ control plane Namespace exists
√ control plane ClusterRoles exist
√ control plane ClusterRoleBindings exist
√ control plane ServiceAccounts exist
√ control plane CustomResourceDefinitions exist
√ control plane MutatingWebhookConfigurations exist
√ control plane ValidatingWebhookConfigurations exist
√ control plane PodSecurityPolicies exist

linkerd-identity
----------------
√ certificate config is valid
√ trust anchors are using supported crypto algorithm
√ trust anchors are within their validity period
√ trust anchors are valid for at least 60 days
√ issuer cert is using supported crypto algorithm
√ issuer cert is within its validity period
√ issuer cert is valid for at least 60 days
√ issuer cert is issued by the trust anchor

linkerd-webhooks-and-apisvc-tls
-------------------------------
√ proxy-injector webhook has valid cert
√ proxy-injector cert is valid for at least 60 days
√ sp-validator webhook has valid cert
√ sp-validator cert is valid for at least 60 days

linkerd-api
-----------
√ control plane pods are ready
√ can initialize the client
√ can query the control plane API

linkerd-version
---------------
√ can determine the latest version
√ cli is up-to-date

control-plane-version
---------------------
√ control plane is up-to-date
√ control plane and cli versions match

linkerd-ha-checks
-----------------
√ pod injection disabled on kube-system

Status check results are √

Linkerd extensions checks
=========================

linkerd-viz
-----------
√ linkerd-viz Namespace exists
√ linkerd-viz ClusterRoles exist
√ linkerd-viz ClusterRoleBindings exist
√ tap API server has valid cert
√ tap API server cert is valid for at least 60 days
√ tap API service is running
√ linkerd-viz pods are injected
√ viz extension pods are running
√ can initialize the client
√ viz extension self-check

Status check results are √

```

### Environment

- Kubernetes Version: v1.18.8
- Cluster Environment: Self hosted, Deployed via kubeadm
- Host OS: Ubuntu 18.04.2 LTS
- Linkerd version: stable-2.10.0

### Possible solution
A workaround option is to use scyllaDB w/o hostNetworking, but this is not desirable as it's recommended to run it with `hostNetwork: true` for increased performance.

**Comment by kleimkuhler** [comment]
@prasus Sorry that you have ran into this issue! We recently closed a duplicate of this: #5881. The PR that closed it was #5904.

This is currently fixed in the most recent `edge-21.3.3` so I'd urge you to try that version. If you prefer to remain on stable, it will also be released as part of `stable-2.10.1`; we don't have an exact date for that but it should be soon.

I'll keep this open if you'd like to confirm this fix on the edge.

**Comment by prasus** [comment]
Thanks much for the update, @kleimkuhler 👍🏿 We have recently upgraded to `stable-2.10.0` and everything seems to work fine except the ones using hostNetworking and we prefer to stay on a stable release so I am closing this for now, will upgrade to `stable-2.10.1` when it's available and open a new request if the problem persists with `stable-2.10.1`, Thanks! 🙏🏿
